internal/global/instruments.go

```diff
@@ -215,9 +215,9 @@ func (i *sfCounter) setDelegate(m metric.Meter) {
 	i.delegate.Store(ctr)
 }
 
-func (i *sfCounter) Add(ctx context.Context, incr float64, attrs ...attribute.KeyValue) {
+func (i *sfCounter) Add(ctx context.Context, incr float64, attrs attribute.Set) {
```

Comment:
Assuming that the proposal is accepted...

Can we do benchmarks what is the difference between accepting a pointer vs value (`attrs *attribute.Set` vs `attrs attribute.Set`)? I see that `attribute.Set` is immutable and it has pointer receivers everywhere so I think it might be intended to be used as a pointer.

I suggest doing benchmarks with different `attrs` payloads (empty set, 3 small attributes, 20 attributes).

Reply:
Important to note here is that [`attribute.NewSet`](https://pkg.go.dev/go.opentelemetry.io/otel@v1.14.0/attribute#NewSet) returns a `Set` not a `*Set` (unlike `attribute.EmptySet`):

Reply:
> Can we do benchmarks what is the difference between accepting a pointer vs value (`attrs *attribute.Set` vs `attrs attribute.Set`)? I see that `attribute.Set` is immutable and it has pointer receivers everywhere so I think it might be intended to be used as a pointer.
> 
> I suggest doing benchmarks with different `attrs` payloads (empty set, 3 small attributes, 20 attributes).

The difference in performance is going to be difference in performance of stack allocations. Goroutine stacks are allocated based on expected size needed (either small (<128 byte), medium (> 128 byte, < 4096 byte), or large (> 4096 byte). Given the stack size of `*Set` is 8 bytes on a 64 bit processor and for `Set` it is 16 (given it's current construction), we can generically test a stack allocation performance:

```go
package main

import (
	"testing"
)

type Data struct {
	// Go stack sizes are defined as:
	//
	//   - small: <= 128
	//   - medium: > 128, < 4096
	//   - big: >= 4096
	buffer [16]byte
}

var dataPtr = &Data{}

func first(d Data) byte { return d.buffer[0] }

func firstPtr(d *Data) byte { return d.buffer[0] }

var result byte

func BenchmarkFirst(b *testing.B) {
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		result = first(*dataPtr)
	}
}

func BenchmarkFirstPtr(b *testing.B) {
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		result = firstPtr(dataPtr)
	}
}
```

```terminal
go test -run='^$' -bench=.
goos: linux
goarch: amd64
pkg: testing/stack
cpu: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
BenchmarkFirst-8      	1000000000	         0.8872 ns/op	       0 B/op	       0 allocs/op
BenchmarkFirstPtr-8   	1000000000	         0.5472 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	testing/stack	1.593s
```

For such small stack sizes there is no real measurable performance difference. The optimization comes for larger struct sizes.

Reply:
More directly, this can be compared with a `Set` if we restructure data:

```go
package main

import (
	"testing"
)

type Data struct {
	buffer interface{}
}

// Go stack sizes are defined as:
//
//   - small: <= 128
//   - medium: > 128, < 4096
//   - big: >= 4096
type buffer [4097]byte

var dataPtr = &Data{buffer: buffer{}}

func first(d Data) byte { return d.buffer.(buffer)[0] }

func firstPtr(d *Data) byte { return d.buffer.(buffer)[0] }

var result byte

func BenchmarkFirst(b *testing.B) {
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		result = first(*dataPtr)
	}
}

func BenchmarkFirstPtr(b *testing.B) {
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		result = firstPtr(dataPtr)
	}
}
```

Still though, the initial evaluation holds:

```terminal
$ go test -run='^$' -bench=.
goos: linux
goarch: amd64
pkg: testing/stack
cpu: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
BenchmarkFirst-8      	 9067317	       124.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkFirstPtr-8   	 9367156	       127.4 ns/op	       0 B/op	       0 allocs/op
PASS
```